### PR TITLE
*: support base get/put.

### PR DESCRIPTION
**Pull request by siddontang**
@ngaut  @disksing

### CODE REVIEW

File: gateway/gateway.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
+	resp, err := kv.Get(ctx, key, opts...)
+	cancel()
+
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if len(resp.Kvs) == 0 {
+		http.NotFound(w, r)
```
Comment (disksing) [inline gateway/gateway.go:131]: I think it's better to write the json format response back to client as etcd API v2 does.

File: gateway/gateway.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+	}
+
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write([]byte(err.Error()))
+		return
+	}
+}
+
+func (gw *Gateway) handleGetKeys(w http.ResponseWriter, r *http.Request) error {
```
Comment (disksing) [inline gateway/gateway.go]: Why naming it `GetKeys` rather than `GetKey`?

File: gateway/gateway.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+	// TODO: parse url args and set there options.
+	var opts []clientv3.OpOption
+
+	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
+	resp, err := kv.Get(ctx, key, opts...)
+	cancel()
+
+	if err != nil {
+		return errors.Trace(err)
```
Comment (disksing) [inline gateway/gateway.go:127]: The error will be send in HTTP body. Maybe Trace will make it hard to read.

File: gateway/gateway.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+	resp, err := gw.client.Txn(ctx).
+		If(cmps...).
+		Then(clientv3.OpPut(key, string(val), opts...)).
+		Commit()
+	cancel()
+
+	if err != nil {
+		return errors.Trace(err)
+	} else if !resp.Succeeded {
+		return errors.Errorf("put value for key %q failed", key)
```
Comment (disksing) [inline gateway/gateway.go:165]: Maybe better to return json format resp.

File: gateway/gateway.go (hunk 2 above)
Comment (siddontang) [inline gateway/gateway.go]: I guess you means `handleKeysGet`?

File: gateway/gateway.go (hunk 1 above)
Comment (siddontang) [inline gateway/gateway.go:131]: no need now.

File: gateway/gateway.go (hunk 3 above)
Comment (siddontang) [inline gateway/gateway.go:127]: doesn't matter. Trace will not change the error message.

File: gateway/gateway.go (hunk 4 above)
Comment (siddontang) [inline gateway/gateway.go:165]: no need now.

**Comment by siddontang** [comment]
PTAL @disksing

**Comment by disksing** [comment]
LGTM

File: gateway/gateway.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+const (
+	etcdTimeout    = 3 * time.Second
+	requestTimeout = 3 * time.Second
+
+	// Etcd use "/v2/keys" for its http v2 prefix, so
+	// here we use /v3/keys.
+	keysPrefix = "/v3/keys"
+
+	contentTypeKey = "Content-Type"
```
Comment (ngaut) [inline gateway/gateway.go:24]: I suppose there is something constant in go's library ?

File: gateway/gateway.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+	if len(resp.Kvs) == 0 {
+		http.NotFound(w, r)
+		return nil
+	}
+
+	writeResponse(w, r, resp)
+	return nil
+}
+
+func (gw *Gateway) handleKeysPut(w http.ResponseWriter, r *http.Request) error {
```
Comment (ngaut) [inline gateway/gateway.go]: s/handleKeysPut/handlePutKeys/g

File: gateway/gateway.go [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+		return errors.Trace(err)
+	}
+
+	w.Header().Set(contentTypeKey, contentType)
+	w.Write(value)
+
+	return nil
+}
+
+func (gw *Gateway) handleKeysGet(w http.ResponseWriter, r *http.Request) error {
```
Comment (ngaut) [inline gateway/gateway.go]: s/handleKeysGet/handleGetKeys/g

File: gateway/gateway_test.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+)
+
+func TestServer(t *testing.T) {
+	TestingT(t)
+}
+
+var _ = Suite(&testGatewaySuite{})
+
+var (
+	addr     = flag.String("addr", "127.0.0.1:8080", "Gateway HTTP listening address")
```
Comment (ngaut) [inline gateway/gateway_test.go]: 8080 is widely used, we may use another port.

File: gateway/gateway.go (hunk 5 above)
Comment (siddontang) [inline gateway/gateway.go:24]: no, go http lib also uses this string directly.

File: gateway/gateway.go (hunk 6 above)
Comment (siddontang) [inline gateway/gateway.go]: This is for Keys handler get method, not get keys.
I will refactor it with a KeysHandler struct instead to prevent confusion.

**Comment by ngaut** [comment]
LGTM

**Pull request merged**